### PR TITLE
Supporting Planes Reservation

### PR DESCRIPTION
**Pull request by Shao-Feng**
The plane of each display used by HWC can be reserved in configuration.
Only the plane assigned in configuration will be used for HWC.

Remaining issue: once the primary plane is not reserved, drmModeAtomicReqPtr
will be NULL from drm.

Change-Id: Ie9a4aa09907cda1920b4735d58c9d9f6d623346a
Tracked-On:https://jira.devtools.intel.com/browse/OAM-71647
Tests: Compile sucessful for Android, feature validation.
Signed-off-by: Shaofeng Tang <shaofeng.tang@intel.com>

### CODE REVIEW

File: common/core/gpudevice.cpp [hunk 1]
```diff
@@ -358,6 +383,50 @@ void GpuDevice::HandleHWCSettings() {
                                  float_rect.at(2), float_rect.at(3));
             float_displays.emplace_back(rect);
           }
+        } else if (!key.compare(key_reserved_drm_plane)) {
+          std::string display_line_str;
```
Comment (js0701) [inline common/core/gpudevice.cpp]: better moved to a separate function

File: hwc_display.ini [hunk 2]
```diff
@@ -43,6 +44,15 @@ FLOAT_DISPLAY="0:320+180+1600+900"
 # cloned-physical-display-number: the display which should clone physical-display-number.
 CLONE_DISPLAY="1+2"
 
+# HWC reserved DRM Plane. 4 DRM planes are available.
+# 0 - Primary Plane
+# 1 - Overlay Plane
+# 2 - Overlay Plane
+# 3 - Cursor Plane
+# 0:0+1+2+3 - ALl plane on display 0 are reserved for HWC
```
Comment (YuanjunHuang) [inline hwc_display.ini]: a little confusing. suggest not to use 'reserve' here. 'All 0/1/2/3 planes of display 0 are used for HWC'

File: hwc_display.ini [hunk 3]
```diff
@@ -43,6 +44,15 @@ FLOAT_DISPLAY="0:320+180+1600+900"
...
+# 1:0+1+3   - Plane 2 will be reserved
```
Comment (YuanjunHuang) [inline hwc_display.ini]: '0/1/3 planes of display 1 are used for HWC, plane 2 is reserved for other component'

File: public/gpudevice.h [hunk 4]
```diff
@@ -103,6 +107,9 @@ class GpuDevice : public HWCThread {
   std::vector<std::unique_ptr<LogicalDisplayManager>> logical_display_manager_;
   std::vector<std::unique_ptr<NativeDisplay>> mosaic_displays_;
   std::vector<NativeDisplay*> total_displays_;
+
+  bool use_reserved_plane_ = false;
```
Comment (YuanjunHuang) [inline public/gpudevice.h]: name is confusing. suggest to use plane_reservation_ or reserve_plane_.

**Comment by Shao-Feng** [comment]
Thanks to Johnson and Yuanjun's comment.
The PR is updated

File: public/gpudevice.h [hunk 5]
```diff
@@ -99,10 +103,14 @@ class GpuDevice : public HWCThread {
...
   void HandleRoutine() override;
   void HandleWait() override;
+  void ParsingPlanReserveSettings(std::string &value);
   std::unique_ptr<DisplayManager> display_manager_;
   std::vector<std::unique_ptr<LogicalDisplayManager>> logical_display_manager_;
   std::vector<std::unique_ptr<NativeDisplay>> mosaic_displays_;
   std::vector<NativeDisplay*> total_displays_;
+
+  bool reserve_plane_ = false;
+  std::map<uint8_t, std::vector<uint32_t>> reserved_drm_plane_map_;
```
Comment (YuanjunHuang) [inline public/gpudevice.h]: suggest to change to reserved_drm_display_planes_map_ to make it more clear

**Comment by Shao-Feng** [comment]
Hotconnect should be supported or not? in this patch, hot-connect is not supported yet

**Comment by Shao-Feng** [comment]
Updated for supporting HotPlug

**Comment by Shao-Feng** [comment]
Add a checking if PlaneReserve is used in HotPlug

**Comment by Shao-Feng** [comment]
The demo is available on https://github.com/Shao-Feng/kmscube

**Comment by Shao-Feng** [comment]
The demo is included in tests/ now

**Comment by js0701** [comment]
the first 2 commit is merged
the demo commit need some IP review
Keep this PR open

**Comment by Shao-Feng** [comment]
README.md and the command option is updated. Atomic is used by default

**Pull request closed without merging**